File: 卡码/代码随想录/1/07-回溯/C++/05-棋盘问题/LeetCode_0051.cpp
//51. N 皇后
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

class Solution {
	vector<vector<string>> res;
	vector<string> chess;
public:
	bool isValid(int row,int col,int n){
		for(int i=0;i<row;i++)
			if(chess[i][col]=='Q') return false;
		for(int i=row-1,j=col-1;i>=0&&j>=0;i--,j--)
			if(chess[i][j]=='Q') return false;
		for(int i=row-1,j=col+1;i>=0&&j<n;i--,j++)
			if(chess[i][j]=='Q') return false;
		return true;
	}
	void backtracking(int row,int n){
		if(row==n){
			res.push_back(chess);
			return;
		}
		for(int col=0;col<n;col++){
			if(isValid(row,col,n)){
				chess[row][col]='Q';
				backtracking(row+1,n);
				chess[row][col]='.';
			}
		}
	}
	vector<vector<string>> solveNQueens(int n) {
		chess.resize(n,string(n,'.'));
		backtracking(0,n);
		return res;
	}
};

int main(){
  system("pause");
  return 0;
}

File: 卡码/代码随想录/1/07-回溯/Js/05-棋盘问题/LeetCode_0051.js
//51. N 皇后
const solveNQueens = function(n) {
  const res=new Array();
  const chess=new Array(n).fill([]).map(()=>new Array(n).fill("."));
  const tmp=new Array(n);
  const isValid=function(row,col){
    for(let i=0;i<row;i++)
      if(chess[i][col]=="Q") return false;
    for(let i=row-1,j=col-1;i>=0&&j>=0;i--,j--)
      if(chess[i][j]=="Q") return false;
    for(let i=row-1,j=col+1;i>=0&&j<n;i--,j++)
      if(chess[i][j]=="Q") return false;
    return true;
  }
  const backtracking=function(row){
    if(row==n){
      chess.forEach((item,index)=>{
        tmp[index]=item.join("");
      });
      res.push(Array.from(tmp));
      return;
    }
    for(let col=0;col<n;col++){
      if(isValid(row,col)){
        chess[row][col]="Q";
        backtracking(row+1);
        chess[row][col]=".";
      }
    }
  }
  backtracking(0);
  return res;
};